packages/graz/src/actions/methods.ts

```diff
@@ -98,3 +99,105 @@ export const sendIbcTokens = async ({
     memo,
   );
 };
+
+export interface InstantiateContractArgs<Message extends Record<string, unknown>> {
+  msg: Message;
+  label: string;
+  fee: StdFee | "auto" | number;
+  options?: InstantiateOptions;
+  senderAddress?: string;
+  codeId: number;
+}
+
+export type InstantiateContractMutationArgs<Message extends Record<string, unknown>> = Omit<
+  InstantiateContractArgs<Message>,
+  "codeId" | "senderAddress" | "fee"
+> & {
+  fee?: StdFee | "auto" | number;
+};
+
+export const instantiateContract = async <Message extends Record<string, unknown>>({
+  senderAddress,
+  msg,
+  fee,
+  options,
+  label,
+  codeId,
+}: InstantiateContractArgs<Message>) => {
+  const { signingClients } = useGrazStore.getState();
+
+  if (!signingClients?.cosmWasm) {
+    throw new Error("Stargate signing client is not ready");
+  }
+  if (!senderAddress) {
+    throw new Error("senderAddress is not defined");
+  }
+
+  return signingClients.cosmWasm.instantiate(senderAddress, codeId, msg, label, fee, options);
+};
+
+export interface ExecuteContractArgs<Message extends Record<string, unknown>> {
+  msg: Message;
+  fee: StdFee | "auto" | number;
+  senderAddress?: string;
+  contractAddress: string;
+}
+
+export type ExecuteContractMutationArgs<Message extends Record<string, unknown>> = Omit<
+  ExecuteContractArgs<Message>,
+  "contractAddress" | "senderAddress" | "fee"
+> & {
+  fee?: StdFee | "auto" | number;
+};
+
+export const executeContract = async <Message extends Record<string, unknown>>({
+  senderAddress,
+  msg,
+  fee,
+  contractAddress,
+}: ExecuteContractArgs<Message>) => {
+  const { signingClients } = useGrazStore.getState();
+
+  if (!signingClients?.cosmWasm) {
+    throw new Error("Stargate signing client is not ready");
+  }
+  if (!senderAddress) {
+    throw new Error("senderAddress is not defined");
+  }
+
+  return signingClients.cosmWasm.execute(senderAddress, contractAddress, msg, fee);
+};
+
+export const getQuerySmart = async <TData>(address?: string, queryMsg?: Record<string, unknown>): Promise<TData> => {
+  const { signingClients } = useGrazStore.getState();
+
+  if (!signingClients?.cosmWasm) {
+    throw new Error("Stargate signing client is not ready");
+  }
+
+  if (address === undefined) {
+    throw new Error("Contract address is undefined");
+  }
+
+  if (queryMsg === undefined) {
+    throw new Error("Query message is undefined");
+  }
+
+  const result = (await signingClients.cosmWasm.queryContractSmart(address, queryMsg)) as TData;
+  return result;
+};
+
+export const getQueryRaw = (keyStr: string, address?: string): Promise<Uint8Array | null> => {
+  const { signingClients } = useGrazStore.getState();
+
+  if (!signingClients?.cosmWasm) {
+    throw new Error("Stargate signing client is not ready");
+  }
+
+  if (address === undefined) {
```

Comment:
Definitely open to this, but wanted to elaborate a bit more on the reasoning just to be be totally clear (and sorry if this is stuff you're aware of!)—

React requires that hooks be instantiated consistently between renders—the same set of hooks must be called in the same order every time the page renders, or React will give a warning and potentially exhibit undefined behavior. This means that if an argument is required, it must be available/defined within the component using the hook from the very first render.

Having optional arguments like this allows us to use the hook in components where the required data (aka contract address or query message) is made available dynamically, rather than having to nest a sub-component that is conditionally rendered when that data is made available. In my experience, it reduces some cruft. Further, it doesn't preclude the user from taking the sub-component approach, either.

The obvious advantage of having everything required is it makes the logic quite a bit simpler and the implementation more straightforward. I'm truly happy to proceed like that, just wanted to share the reasoning first!

Reply:
The primary use case within `react-query` is [dependent queries](https://tanstack.com/query/v4/docs/guides/dependent-queries)

Reply:
Yes we can still keep that behavior in hooks level, by using `enabled` configuration. in action level we can make them required

example:
```
const getSomething = (foo:string, bar:string) => {
// foo and bar required in fetchSomething
 return await fetchSomething(foo, bar)
}

const useSomething = ({foo, bar}:{foo?: string, bar?: string}) => {
  const query = useQuery(["USE_SOMETHING", foo, bar], async () => {
  return await getSomething(foo, bar)
}, {
  enabled: Boolean(foo) && Boolean(bar)
})
  
}
```

hooks still can be called when `foo` & `bar` undefined but will execute if `foo` & `bar` is there and `foo` & `bar` changed

Reply:
Seems like a more than reasonable approach :)